src/key.rs

```diff
@@ -891,7 +891,7 @@ impl Keypair {
         match from_hex(s, &mut res) {
             Ok(constants::SECRET_KEY_SIZE) =>
                 Keypair::from_seckey_slice(secp, &res[0..constants::SECRET_KEY_SIZE]),
-            _ => Err(Error::InvalidPublicKey),
+            _ => Err(Error::InvalidSecretKey),
```

Comment:
Oh yes, the entire error type needs a lot of work...

Reply:
I think we should move deserialization/serialization into Rust and then we can provide sensible error types. Obviously in this case the variant was totally wrong, but in general the error types in this library are crappy because they're just thin wrappers around the numeric error returns from the C code.

Reply:
...lol, and half this API is "take a `&[u8]` then length-check it" and should be replaced by an array-based on.

Reply:
Ye, I think we need to do all checks except "is on the curve"

Reply:
I remember that there was an initiative to replace slice-based methods with array-based ones. IDK if this was Miniscript or secp. In any case, I'm happy to write a follow-up PR to move the constructors to array parameters.

Reply:
@uncomputable that would be awesome.

If you want a fun weekend project you can also try to pull some parts out of https://github.com/rust-bitcoin/rust-secp256k1/pull/703 to replace point deserialization code with Rust. (That PR is out of sync and also IMO it does too much at once; we should practice Rust-ifying a couple functions and defining sensible APIs and error types.)